qiskit/transpiler/passes/utils/gates_basis.py

```diff
@@ -0,0 +1,39 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2017, 2018.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Check if a property reached a fixed point."""
+
+from qiskit.transpiler.basepasses import AnalysisPass
+
+
+class GatesInBasis(AnalysisPass):
+    """Check if a a dag is all in the basis"""
+
+    def __init__(self, basis_gates):
+        """FixedPoint initializer.
+
+        Args:
+            basis_gates (list): The list of basis gate names to check
+        """
+        super().__init__()
+        self._basis_gates = set(basis_gates)
+
+    def run(self, dag):
+        """Run the GatesInBasis pass on `dag`."""
+        gates_out_of_basis = False
+        gates = set(x.op.name for x in dag.gate_nodes())
+        for gate in gates:
+            if gate not in self._basis_gates:
+                gates_out_of_basis = True
+                break
```

Comment:
Following up on our performance discussion, compared to the alternative (checking gate-wise `if x.op.name not in self._basis_gates`) this might miss an opportunity to exit early and avoid scanning the full circuit if there was an out-of-basis gate, but given that `dag.gate_nodes()` will scan the full circuit regardless, and by this point in the optimization loop, its more likely that most/all of the gates will be in the basis, this seems the best approach.

Reply:
It looks like `dag.gate_nodes()` ultimately constructs a few lists. It might be more generally useful to offer an iterator over nodes in `rx.PyDAG()`

Reply:
I just read several issues related to python access to rust iterators and see that it is not possible in general: https://github.com/Qiskit/retworkx/issues/71
Instead, it is possible to construct python iterators over rust sequences, like `Vec`: https://github.com/Qiskit/retworkx/pull/185 @mtreinish pointed out that this avoids converting the entire sequence to a python list, which is often the most expensive part. If it turns out that exiting early from the loop might give a performance gain, then using this kind of iterator might help.

> I would've expected though the BasisTranslator to effectively be a no-op if the circuit is already in the basis (and effectively have a version of this check implemented already).

It does check

https://github.com/Qiskit/qiskit-terra/blob/8d5bcca2003b7033cd32087286918e72d4421d34/qiskit/transpiler/passes/basis/basis_translator.py#L130-L132

This appears to have the same efficiency as `GatesInBasis`.  But, there is quite a bit of work done before this loop is entered, including another call `dag.op_nodes`.